# First load the modules
#        module load snakemake bwa-mem
#
# usage: $ snakemake -np ../sorted_reads/B.bam
#          snakemake -np ../mapped_reads/A.bam ../mapped_reads/B.bam
#          snakemake -np ../mapped_reads/{A,B}.bam


rule bwa_map:
    input:
        "../data/genome.fa",
        "../data/samples/{sample}.fastq"
    output:
        "../mapped_reads/{sample}.bam"
    shell:
        "bwa mem {input} | samtools view -Sb - > {output}"
